Add support for mirroring the current view to a different swap chain.

tpsiaki opened the pull request:
This adds support for saving/replaying the current viewRenderTarget which is helpful for mirroring a rendered view onto multiple surfaces.

romainguy commented:
It looks fine but the terminology is a bit off. It's not really a replay.

tpsiaki commented:
Yeah agreed that replay isn't right.  I was trying to think of something better but blitSavedFrame didn't seem right.  How about something like storeFrame/loadFrame?

romainguy commented:
Does it have to be store/load? This would produce 2 blits, but we should be able to do only one. We can blit directly from our framebuffer to the mirror framebuffer.

tpsiaki commented:
I'm not sure how we can avoid two blits here.  The postprocess isn't complete until the final blit into the default framebuffer when the main rendered swap chain is active, and so we have to bind that as the read framebuffer, and then we need to blit this into the default framebuffer when the mirrored swap chain is active, so I'm not sure how we can do that without blitting to an intermediate named render target.  i.e. I don't think two different gl displays' default render targets can be bound simultaneously to read/draw for a blit.

I considered just having a single mirror call that did a blit to a named render target from the pool, then made a new swap chain active, then blitted the named render target to the default framebuffer of the new swap chain, then made the previous swap chain active again, but it seemed kind of wrong to switch to a different context/swap chain in the middle of rendering a frame.

romainguy commented:
You can set one eglSurface as the read buffer and another one as the draw buffer. Grafika has an example of this I believe.

tpsiaki commented:
Thx.  grafika ftw!  That worked.  New changes uploaded that replace the save/load with a single mirror call, and avoid the double blit.

pixelflinger commented on filament/include/filament/Renderer.h:
```diff
@@ -136,6 +136,20 @@ class UTILS_PUBLIC Renderer : public FilamentAPI {
+     * mirrorFrame() should be called after a frme is rendered using render()
+     * but before endFrame() is called.
+     */
+    void mirrorFrame(SwapChain* toSwapChain, View const* toView, View const* fromView);
```
The parameters shouldn't be `View*` but `Viewport*`. Also use `dst` and `src`.

tpsiaki replied:
done.

pixelflinger commented on filament/src/Renderer.cpp:
```diff
@@ -235,6 +235,27 @@ void FRenderer::renderJob(ArenaScope& arena, FView* view) {
+    assert(toSwapChain);
+    FEngine& engine = getEngine();
+    FEngine::DriverApi& driver = engine.getDriverApi();
+    Viewport const& fromVp = fromView->getViewport();
```
use `src`/`dst` instead of `from`/`to`

tpsiaki replied:
done

tpsiaki commented on filament/src/Renderer.cpp:
```diff
@@ -235,6 +235,27 @@ void FRenderer::renderJob(ArenaScope& arena, FView* view) {
+    FEngine& engine = getEngine();
+    FEngine::DriverApi& driver = engine.getDriverApi();
+    Viewport const& fromVp = fromView->getViewport();
+    Viewport const& toVp = toView->getViewport();
```
done.

tpsiaki commented on filament/src/Renderer.cpp:
```diff
@@ -235,6 +235,27 @@ void FRenderer::renderJob(ArenaScope& arena, FView* view) {
+
+    const Handle<HwRenderTarget> viewRenderTarget = getRenderTarget();
+
+    toSwapChain->makeCurrent(driver, mSwapChain);
```
done.

tpsiaki commented on filament/src/details/Renderer.h:
```diff
@@ -139,6 +141,7 @@ class FRenderer : public Renderer {
     FrameInfoManager mFrameInfoManager;
     bool mIsRGB16FSupported : 1;
     bool mIsRGB8Supported : 1;
+    RenderTargetPool::Target const* savedFrameTarget = nullptr;
```
Removed.  this was leftover from a previous iteration.

tpsiaki commented on filament/src/details/SwapChain.h:
```diff
@@ -36,7 +36,11 @@ class FSwapChain : public SwapChain {
+        driverApi.makeCurrent(mSwapChain, mSwapChain);
+    }
+
+    void makeCurrent(driver::DriverApi& driverApi, FSwapChain* readSwapChain) noexcept {
```
Done.  I had to add a getHwHandle api to SwapChain to make this work, but that seemed to be consistent with other things (e.g. RenderPrimitive)

pixelflinger commented on filament/src/Renderer.cpp:
```diff
@@ -235,24 +235,29 @@ void FRenderer::renderJob(ArenaScope& arena, FView* view) {
+                viewRenderTarget, srcViewport->left, srcViewport->bottom, srcViewport->width, srcViewport->height);
+    dstSwapChain->commit(driver);
 
+    // Reset the context and read/draw surface4w to the current surface so that
```
typo: `surface4w`

tpsiaki replied:
done.

tpsiaki commented:
@pixelflinger Regarding swap chain flags - I added a READABLE config for swap chain creation, and I added an assert in the Renderer.mirrorFrame call to make sure that the src SwapChain was readable.  Is that the extent of what you were looking for, or is there something additional that needs to happen when creating the Vulkan swap chain to respect that flag?

pixelflinger commented on filament/src/Renderer.cpp:
```diff
@@ -235,24 +235,29 @@ void FRenderer::renderJob(ArenaScope& arena, FView* view) {
-    toSwapChain->commit(driver);
+                viewRenderTarget, dstViewport->left, dstViewport->bottom, dstViewport->width, dstViewport->height,
+                viewRenderTarget, srcViewport->left, srcViewport->bottom, srcViewport->width, srcViewport->height);
+    dstSwapChain->commit(driver);
```
Do we really want to do a `commit()` here? The destination swapchain is not owned by filament. A commit is a big deal -- what if the caller wanted to do additional rendering? Conceptually that swapchain could be owned by someone else, maybe even another filament instance?

I feel like it should be the responsibility of the caller to do this. This method should be just about drawing into it.

Moreover, how does this interact with `setPresentationTime()`? I think it doesn't and that's a problem. The presentation time should be set on that surface -- and again I'm not sure it is filament's duty to do that.

I'm okay if we decide to go with the `commit()` here, but it should be clearly documented in the doxygen doc.

tpsiaki replied:
added flags to control whether commit and setPresentationTime are called on dstSwapChain

pixelflinger commented on filament/include/filament/Renderer.h:
```diff
@@ -140,15 +141,15 @@ class UTILS_PUBLIC Renderer : public FilamentAPI {
      * but before endFrame() is called.
      */
-    void mirrorFrame(SwapChain* toSwapChain, View const* toView, View const* fromView);
+    void mirrorFrame(SwapChain* dstSwapChain, Viewport const* dstViewport, Viewport const* srcViewport);
```
Ah sorry for not asking this before, please use `const&` instead of `const*` for consistency with the rest of the API.

tpsiaki replied:
done.

tpsiaki commented:
Added flags for setting presentation time & commit, and switched viewports to pass by const ref.

pixelflinger commented on filament/src/details/SwapChain.h:
```diff
@@ -51,6 +51,14 @@ class FSwapChain : public SwapChain {
     }
 
+    constexpr bool isReadable() const noexcept {
+        return (mConfigFlags * CONFIG_READABLE) != 0;
```
typo: should be `&`

tpsiaki replied:
done.

pixelflinger commented on filament/include/filament/SwapChain.h:
```diff
@@ -143,6 +143,7 @@ namespace filament {
 class UTILS_PUBLIC SwapChain : public FilamentAPI {
 public:
     static const uint64_t CONFIG_TRANSPARENT = driver::SWAP_CHAIN_CONFIG_TRANSPARENT;
+    static const uint64_t CONFIG_READABLE = driver::SWAP_CHAIN_CONFIG_READABLE;
```
There should be some documentation for this new flag.

tpsiaki replied:
done.

pixelflinger commented on filament/src/Renderer.cpp:
```diff
@@ -235,6 +235,40 @@ void FRenderer::renderJob(ArenaScope& arena, FView* view) {
+    // destination.
+    driver.makeCurrent(dstSwapChain->getHwHandle(), mSwapChain->getHwHandle());
+
+    driver.blit(TargetBufferFlags::COLOR,
```
Check that the source SwapChain has the IS_READABLE flag, so we can ensure clients use this properly. I would use an assert() at least.

tpsiaki replied:
it was earlier in the function.  Moved it right before the blit and added a comment.
